### PR TITLE
Add the C++ gradient of the Prod operation.

### PR DESCRIPTION
**Pull request by theflofly**
This PR adds the gradient of the Prod operation to the C++ API.

### CODE REVIEW

File: tensorflow/cc/gradients/math_grad.cc [hunk 1]
```diff
@@ -794,6 +794,183 @@ Status MinOrMaxGrad(const Scope& scope, const Operation& op,
...
+  auto grad_tiled = Tile(scope, grad, tile_scaling);
+
+  Scope cpu_scope = scope.WithDevice("/cpu:0");
+
+  // [3]
+  auto rank = Rank(cpu_scope, op.input(0));
+
+
+  // Normalize any negative indices in the reduction_axes to positive values.
+  auto reduction_indices_pos = Mod(scope, Add(scope, reduction_indices, rank), rank);
```
Comment (alextp) [inline tensorflow/cc/gradients/math_grad.cc]: If this is intended as a line-by-line translation of the python code these operations should be in the cpu_scope as well and not in scope.

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/cc/gradients/math_grad.cc (hunk 1 above)
Comment (theflofly) [reply tensorflow/cc/gradients/math_grad.cc]: My bad, I added it afterward. I updated.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by theflofly** [comment]
@alextp @drpngx Any news regarding this PR?

**Review by alextp — APPROVED** [review]

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by theflofly** [comment]
I merged with master in order to resolve conflicts. It's been more than a month since this PR has been submitted...

**Comment by yifeif** [comment]
Thank you @theflofly. Once the tests pass, we can merge the PR asap!

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by theflofly** [comment]
Thanks @yifeif !

**Pull request merged**